gui/packages/desktop/src/main/index.js

```diff
@@ -157,24 +157,48 @@ const ApplicationMain = {
     }
   },
 
-  _onBeforeQuit() {
-    this._shouldQuit = true;
+  async _onBeforeQuit(event: Event) {
+    switch (this._quitStage) {
+      case 'unready':
+        // postpone the app shutdown
+        event.preventDefault();
 
-    if (process.env.NODE_ENV === 'development') {
-      if (this._windowController) {
-        this._windowController.window.closeDevTools();
-      }
+        this._quitStage = 'initiated';
+        await this._prepareToQuit();
+
+        // terminate the app
+        this._quitStage = 'ready';
+        app.quit();
+        break;
+
+      case 'initiated':
+        // prevent immediate exit, the app will quit after running the shutdown routine
+        event.preventDefault();
+        return;
+
+      case 'ready':
+        // let the app quit freely at this point
+        break;
     }
+  },
 
-    return true;
+  async _prepareToQuit() {
+    if (this._connectedToDaemon) {
```

Comment:
FYI Slight improvement, no need to wait for RPC client to fail before terminating the app in extreme case when the daemon is not running.